Add ReplicatedParameter and ReplicatedTensor to DDP.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74787

As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

pritamdamania committed:
Add ReplicatedParameter and ReplicatedTensor to DDP.

As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/replicated_tensor.py:
```diff
@@ -123,3 +123,27 @@ def validate(self) -> bool:
 
         return True
+
+class ReplicatedParameter(ReplicatedTensor, torch.nn.Parameter):
```
@albanD I couldn't pass in a ReplicatedTensor to `torch.nn.Parameter` and have `isinstance()` still return ReplicatedTensor. As a result, I had to do something messy like this. Not sure if there is a better way to do this.

albanD replied:
cc @jbschlosser

jbschlosser replied:
This is a bit delayed due to my medical leave, but wanted to leave an update here for posterity: #73459 is in now - please try it out if you'd like :)

mrshenli reviewed (commented):
Is this a test or do we plan to land this soon? If it's the latter, shall we do some micro-benchmarks to make sure there is no perf regression?

kumpera commented on torch/nn/parallel/distributed.py:
```diff
@@ -665,6 +665,16 @@ def _sync_params_and_buffers(self, authoritative_rank=0):
             )
 
+        # After syncing, tag them as replicated.
+        from torch.distributed._shard.replicated_tensor import ReplicatedTensor, ReplicatedParameter
```
Shouldn't we check if each `param` or `buffer` is replicated before converting them?
Similarly, shouldn't we record which ones we wrapped and use that when unwrapping, instead of relying on `parameters_to_ignore`.

Another thing, is using `parameters_to_ignore` to filter buffers just a naming artifact?

pritamdamania87 replied:
> Similarly, shouldn't we record which ones we wrapped and use that when unwrapping, instead of relying on parameters_to_ignore.

So `parameters_to_ignore` is slightly orthogonal to tagging something as replicated or not since that option is actually controlling which parameters DDP ignores completely. That is why we always need to rely on it and ensure we don't tag `parameters_to_ignore` as Replicated at all.

> Another thing, is using parameters_to_ignore to filter buffers just a naming artifact?

It's an unfortunate name since it is filtering buffers as well in other places. I'll probably fix this name in a follow up PR.

rohan-varma replied:
Oops, I added `self.parameters_to_ignore` a couple of years ago and agree that a better name is `parameters_and_buffers_to_ignore`.

pritamdamania committed:
Update on "Add ReplicatedParameter and ReplicatedTensor to DDP."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

[ghstack-poisoned]

pritamdamania87 commented:
@mrshenli The plan is to hopefully land this change, however I think there are some tricky cases to take care of first. I'll probably ensure I can make CI happy first (tagged the PR as WIP) and then try a benchmark as well to ensure no regressions.

pritamdamania committed:
Update on "Add ReplicatedParameter and ReplicatedTensor to DDP."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot retest this please

pritamdamania committed:
Update on "Add ReplicatedParameter and ReplicatedTensor to DDP."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add ReplicatedParameter and ReplicatedTensor to DDP."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add ReplicatedParameter and ReplicatedTensor to DDP."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add ReplicatedParameter and ReplicatedTensor to DDP."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -74,3 +76,71 @@ def test_replicated_tensor_inter_op_tensor(self):
+            self.assertIsInstance(out, ReplicatedTensor)
+
+            # Validate buffers not replicated after forward.
+            self.assertNotIsInstance(ddp.module.foo, ReplicatedTensor)
```
Don't we synchronize the buffers so it should be replicated?

rohan-varma commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -74,3 +76,71 @@ def test_replicated_tensor_inter_op_tensor(self):
+        params = list(model.parameters())
+        self.assertEqual(2, len(params))
+        self.assertEqual(model.weight, params[0])
+        self.assertEqual(model.bias, params[1])
```
should we add `self.assertTrue(all(isinstance(replicatedParam, param) for param in model.parameters())`?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -601,11 +601,14 @@ def __init__(
         self.gradient_as_bucket_view = gradient_as_bucket_view
-        if hasattr(module, "_ddp_params_and_buffers_to_ignore"):
-            self.parameters_to_ignore = module._ddp_params_and_buffers_to_ignore
+        if hasattr(self.module, "_ddp_params_and_buffers_to_ignore"):
```
Curious why this change is needed?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -637,10 +640,10 @@ def __init__(
         # In debug mode, build a mapping of parameter index -> parameter.
         param_to_name_mapping = self._build_debug_param_to_name_mapping(parameters)
+        # Sync params and buffers. Ensures all DDP models start off at the same value.
+        self._sync_params_and_buffers(authoritative_rank=0)
```
Shall we only mark this after the sync completes successfully?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -665,6 +668,27 @@ def _sync_params_and_buffers(self, authoritative_rank=0):
+            if not buffers_only:
+                for param_name, param in module.named_parameters(recurse=False):
+                    if (param.requires_grad and f'{module_name}.{param_name}'
+                            not in self.parameters_to_ignore and not isinstance(param, ReplicatedParameter)):
```
does the latter check mean that it is possible to call this function on already replicated parameters? Should we try to make it so that the application doesn't do this because it would be unnecessary?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -665,6 +668,27 @@ def _sync_params_and_buffers(self, authoritative_rank=0):
+                for param_name, param in module.named_parameters(recurse=False):
+                    if (param.requires_grad and f'{module_name}.{param_name}'
+                            not in self.parameters_to_ignore and not isinstance(param, ReplicatedParameter)):
+                        module.register_parameter(param_name, ReplicatedParameter(param))
```
Do we not de-register the original, non-replicated parameters?

rohan-varma reviewed (commented):
Are we considering enabling this change with a flag such as `DDP_USE_REPLICATED` to phase the rollout? I would advocate for this unless we're very confident about performance implications and reliability of `ReplicatedTensor`, as it would also have a large blast radius in OSS, and we may not be completely aware of it until the 1.12 release happens.

pritamdamania87 commented:
Closing in favor of: https://github.com/pytorch/pytorch/pull/75299
